src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,163 @@
+        if (index == 0) {
+            current.next.prev = node;
+            node.prev = null;
+            node.next = current.next;
+            head = node;
+        } else {
+            node.next = current.next;
+            node.prev = current.prev;
+            current.prev.next = node;
+            current.next.prev = node;
+        }
+        return current.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        throwIndexOutOfBounds(index);
+        Node<T> current = findNodeByIndex(index);
+        if (current.equals(head)) {
+            head = head.next;
+        } else {
+            current.prev.next = current.next;
+        }
+        if (current.equals(tail)) {
+            tail = tail.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.item == object || (current.item != null && current.item.equals(object))) {
+                if (current == head) {
+                    head = current.next;
+                }
+                if (current == tail) {
+                    tail = current.prev;
+                }
+                if (current.prev != null) {
+                    current.prev.next = current.next;
+                }
+                if (current.next != null) {
+                    current.next.prev = current.prev;
+                }
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private final T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+
+        public Node(T item) {
+            this.item = item;
+        }
+    }
+
+    private void throwIndexOutOfBoundsAdd(int index) throws IndexOutOfBoundsException {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds.");
+        }
+    }
+
+    private void throwIndexOutOfBounds(int index) throws IndexOutOfBoundsException {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds.");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        throwIndexOutOfBounds(index);
+        Node<T> node;
+        node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
```

Comment:
The `findNodeByIndex(int index)` method can be optimized by iterating only the first or second half of the list, depending on the index value. That will boost your lists' performance.